File: server/mcp/character_normalizer.py
"""Character normalizer — post-generation silhouette detection and normalization.

Ported from characterNormalizer.js. Detects character silhouette in
generated images, normalizes scale/position/anchor, detects head region
for face mask protection during refinement passes.
"""

from __future__ import annotations

import logging
from dataclasses import dataclass

from PIL import Image

logger = logging.getLogger(__name__)

WHITE_THRESHOLD = 240
MIN_PIXEL_RATIO = 0.005


@dataclass(frozen=True)
class Region:
    x: float
    y: float
    w: float
    h: float


@dataclass(frozen=True)
class Point:
    x: float
    y: float


@dataclass(frozen=True)
class Silhouette:
    bounds: Region
    center: Point
    pixel_ratio: float
    width_px: int
    height_px: int


@dataclass(frozen=True)
class HeadRegion:
    bounds: Region
    center: Point
    radius: float


@dataclass(frozen=True)
class Normalization:
    scale: float
    offset_x: float
    offset_y: float
    alignment_score: float | None
    expected_bounds: Region
    actual_bounds: Region
    needs_correction: bool


def _is_white_or_near(r: int, g: int, b: int, a: int) -> bool:
    if a < 128:
        return True
    return r > WHITE_THRESHOLD and g > WHITE_THRESHOLD and b > WHITE_THRESHOLD


def detect_silhouette(image: Image.Image) -> Silhouette | None:
    img = image.convert("RGBA")
    w, h = img.size
    pixels = img.get_flattened_data()

    min_x, min_y = w, h
    max_x = max_y = 0
    pixel_count = 0

    for y in range(h):
        for x in range(w):
            i = y * w + x
            r, g, b, a = pixels[i]
            if not _is_white_or_near(r, g, b, a):
                if x < min_x:
                    min_x = x
                if x > max_x:
                    max_x = x
                if y < min_y:
                    min_y = y
                if y > max_y:
                    max_y = y
                pixel_count += 1

    if pixel_count < w * h * MIN_PIXEL_RATIO:
        return None

    return Silhouette(
        bounds=Region(
            x=min_x / w,
            y=min_y / h,
            w=(max_x - min_x) / w,
            h=(max_y - min_y) / h,
        ),
        center=Point(
            x=(min_x + max_x) / 2 / w,
            y=(min_y + max_y) / 2 / h,
        ),
        pixel_ratio=pixel_count / (w * h),
        width_px=max_x - min_x,
        height_px=max_y - min_y,
    )


def detect_head_region(
    image: Image.Image,
    silhouette: Silhouette,
) -> HeadRegion | None:
    img = image.convert("RGBA")
    w, h = img.size
    pixels = img.get_flattened_data()

    top_y = silhouette.bounds.y * h
    center_x = silhouette.center.x * w
    search_height = silhouette.bounds.h * h * 0.3
    search_width = silhouette.bounds.w * w * 0.4

    min_x, min_y = w, h
    max_x = max_y = 0
    count = 0

    start_y = max(0, int(top_y) - 5)
    end_y = min(h, int(top_y + search_height))
    start_x = max(0, int(center_x - search_width / 2))
    end_x = min(w, int(center_x + search_width / 2))

    for y in range(start_y, end_y):
        for x in range(start_x, end_x):
            i = y * w + x
            r, g, b, a = pixels[i]
            if not _is_white_or_near(r, g, b, a):
                if x < min_x:
                    min_x = x
                if x > max_x:
                    max_x = x
                if y < min_y:
                    min_y = y
                if y > max_y:
                    max_y = y
                count += 1

    if count < 50:
        return None

    return HeadRegion(
        bounds=Region(
            x=min_x / w,
            y=min_y / h,
            w=(max_x - min_x) / w,
            h=(max_y - min_y) / h,
        ),
        center=Point(
            x=(min_x + max_x) / 2 / w,
            y=(min_y + max_y) / 2 / h,
        ),
        radius=max(max_x - min_x, max_y - min_y) / 2 / w,
    )


def compute_normalization(
    silhouette: Silhouette,
    head_region: HeadRegion | None,
    pose_bounds: Region | None,
    canvas_size: int = 1024,
) -> Normalization | None:
    if not silhouette or not pose_bounds:
        return None

    expected_h = pose_bounds.h * canvas_size
    expected_w = pose_bounds.w * canvas_size

    actual_h = silhouette.height_px
    actual_w = silhouette.width_px

    scale_y = expected_h / actual_h
    scale_x = expected_w / actual_w
    scale = min(scale_x, scale_y)

    expected_anchor_x = pose_bounds.x * canvas_size + expected_w / 2
    expected_anchor_y = (pose_bounds.y + pose_bounds.h) * canvas_size

    actual_anchor_x = silhouette.bounds.x * canvas_size + actual_w * scale / 2
    actual_anchor_y = (silhouette.bounds.y + silhouette.bounds.h) * canvas_size * scale

    offset_x = expected_anchor_x - actual_anchor_x
    offset_y = expected_anchor_y - actual_anchor_y

    alignment_score = None
    if head_region and pose_bounds:
        alignment_score = 1 - min(
            1.0,
            abs(head_region.center.x - pose_bounds.x - pose_bounds.w / 2)
            + abs(head_region.center.y - pose_bounds.y - pose_bounds.h * 0.15),
        )

    needs_correction = (
        scale < 0.8 or scale > 1.25 or (alignment_score is not None and alignment_score < 0.7)
    )

    return Normalization(
        scale=scale,
        offset_x=offset_x,
        offset_y=offset_y,
        alignment_score=alignment_score,
        expected_bounds=pose_bounds,
        actual_bounds=silhouette.bounds,
        needs_correction=needs_correction,
    )


def normalize_character(
    image: Image.Image,
    pose_bounds: Region | None = None,
    target_width: int = 1024,
    target_height: int = 1024,
) -> tuple[Image.Image, Silhouette | None, HeadRegion | None, Normalization | None]:
    src = image.convert("RGBA")
    silhouette = detect_silhouette(src)
    if not silhouette:
        return (
            src.resize((target_width, target_height), Image.Resampling.LANCZOS),
            None,
            None,
            None,
        )

    head_region = detect_head_region(src, silhouette)
    norm = compute_normalization(silhouette, head_region, pose_bounds, target_width)

    out = Image.new("RGBA", (target_width, target_height), (255, 255, 255, 255))

    if norm:
        sx = int(silhouette.bounds.x * src.width)
        sy = int(silhouette.bounds.y * src.height)
        sw = silhouette.width_px
        sh = silhouette.height_px
        dw = int(sw * norm.scale)
        dh = int(sh * norm.scale)
        dx = int(norm.offset_x + silhouette.bounds.x * target_width * norm.scale)
        dy = int(norm.offset_y + silhouette.bounds.y * target_height * norm.scale)
        crop = src.crop((sx, sy, sx + sw, sy + sh))
        out.paste(crop.resize((dw, dh), Image.Resampling.LANCZOS), (dx, dy))
    else:
        scale = min(
            target_width * 0.6 / silhouette.width_px,
            target_height * 0.9 / silhouette.height_px,
        )
        dw = int(silhouette.width_px * scale)
        dh = int(silhouette.height_px * scale)
        dx = (target_width - dw) // 2
        dy = int(target_height * 0.95) - dh
        sx = int(silhouette.bounds.x * src.width)
        sy = int(silhouette.bounds.y * src.height)
        crop = src.crop((sx, sy, sx + silhouette.width_px, sy + silhouette.height_px))
        out.paste(crop.resize((dw, dh), Image.Resampling.LANCZOS), (dx, dy))

    return (out, silhouette, head_region, norm)


def build_face_mask(
    head_region: HeadRegion | None,
    canvas_size: int = 1024,
) -> Image.Image | None:
    if not head_region:
        return None

    from PIL import ImageDraw

    mask = Image.new("L", (canvas_size, canvas_size), 0)
    draw = ImageDraw.Draw(mask)

    cx = head_region.center.x * canvas_size
    cy = head_region.center.y * canvas_size
    rx = (head_region.bounds.w * canvas_size) / 2 * 1.2
    ry = (head_region.bounds.h * canvas_size) / 2 * 1.1

    draw.ellipse(
        [cx - rx, cy - ry, cx + rx, cy + ry],
        fill=255,
    )
    return mask

File: src/lib/characterNormalizer.js
// Character Normalizer — post-generation correction for pose-locking.
// Detects character silhouette in generated image, normalizes scale/position/anchor.

const WHITE_THRESHOLD = 240;
const MIN_PIXEL_RATIO = 0.005;

function isWhiteOrNear(r, g, b, a) {
  if (a < 128) return true;
  return r > WHITE_THRESHOLD && g > WHITE_THRESHOLD && b > WHITE_THRESHOLD;
}

export function detectSilhouette(imageData, width, height) {
  const data = imageData.data;
  let minX = width, minY = height, maxX = 0, maxY = 0;
  let pixelCount = 0;

  for (let y = 0; y < height; y++) {
    for (let x = 0; x < width; x++) {
      const i = (y * width + x) * 4;
      if (!isWhiteOrNear(data[i], data[i + 1], data[i + 2], data[i + 3])) {
        if (x < minX) minX = x;
        if (x > maxX) maxX = x;
        if (y < minY) minY = y;
        if (y > maxY) maxY = y;
        pixelCount++;
      }
    }
  }

  if (pixelCount < width * height * MIN_PIXEL_RATIO) {
    return null;
  }

  return {
    bounds: {
      x: minX / width,
      y: minY / height,
      w: (maxX - minX) / width,
      h: (maxY - minY) / height,
    },
    center: {
      x: (minX + maxX) / 2 / width,
      y: (minY + maxY) / 2 / height,
    },
    pixel_ratio: pixelCount / (width * height),
    width_px: maxX - minX,
    height_px: maxY - minY,
  };
}

export function detectHeadRegion(imageData, width, height, silhouette) {
  if (!silhouette) return null;
  const data = imageData.data;
  const topY = silhouette.bounds.y * height;
  const centerY = silhouette.center.x * width;
  const searchHeight = silhouette.bounds.h * height * 0.3;
  const searchWidth = silhouette.bounds.w * width * 0.4;

  let minX = width, minY = height, maxX = 0, maxY = 0;
  let count = 0;

  const startY = Math.floor(Math.max(0, topY - 5));
  const endY = Math.floor(Math.min(height, topY + searchHeight));
  const startX = Math.floor(Math.max(0, centerY - searchWidth / 2));
  const endX = Math.floor(Math.min(width, centerY + searchWidth / 2));

  for (let y = startY; y < endY; y++) {
    for (let x = startX; x < endX; x++) {
      const i = (y * width + x) * 4;
      if (!isWhiteOrNear(data[i], data[i + 1], data[i + 2], data[i + 3])) {
        if (x < minX) minX = x;
        if (x > maxX) maxX = x;
        if (y < minY) minY = y;
        if (y > maxY) maxY = y;
        count++;
      }
    }
  }

  if (count < 50) return null;

  return {
    bounds: {
      x: minX / width,
      y: minY / height,
      w: (maxX - minX) / width,
      h: (maxY - minY) / height,
    },
    center: {
      x: (minX + maxX) / 2 / width,
      y: (minY + maxY) / 2 / height,
    },
    radius: Math.max(maxX - minX, maxY - minY) / 2 / width,
  };
}

export function computeNormalization(silhouette, headRegion, poseGeo, canvasSize = 1024) {
  if (!silhouette || !poseGeo) return null;

  const expectedBounds = poseGeo.bounds;
  const expectedH = expectedBounds.h * canvasSize;
  const expectedW = expectedBounds.w * canvasSize;

  const actualH = silhouette.height_px;
  const actualW = silhouette.width_px;

  const scaleY = expectedH / actualH;
  const scaleX = expectedW / actualW;
  const scale = Math.min(scaleX, scaleY);

  const expectedAnchorX = expectedBounds.x * canvasSize + expectedW / 2;
  const expectedAnchorY = (poseGeo.ground_y || expectedBounds.y + expectedBounds.h) * canvasSize;

  const actualAnchorX = silhouette.bounds.x * canvasSize + actualW * scale / 2;
  const actualAnchorY = (silhouette.bounds.y + silhouette.bounds.h) * canvasSize * scale;

  const offsetX = expectedAnchorX - actualAnchorX;
  const offsetY = expectedAnchorY - actualAnchorY;

  const alignmentScore = headRegion && poseGeo.head_center
    ? 1 - Math.min(1, Math.abs(headRegion.center.x - poseGeo.head_center.x) +
                     Math.abs(headRegion.center.y - poseGeo.head_center.y))
    : null;

  return {
    scale,
    offset: { x: offsetX, y: offsetY },
    alignment_score: alignmentScore,
    expected_bounds: expectedBounds,
    actual_bounds: silhouette.bounds,
    needs_correction: scale < 0.8 || scale > 1.25 || (alignmentScore !== null && alignmentScore < 0.7),
  };
}

export async function normalizeCharacter(imageUrl, poseGeo, targetWidth = 1024, targetHeight = 1024) {
  return new Promise((resolve) => {
    const img = new Image();
    img.crossOrigin = "anonymous";
    img.onload = () => {
      const srcCanvas = document.createElement("canvas");
      srcCanvas.width = img.naturalWidth;
      srcCanvas.height = img.naturalHeight;
      const srcCtx = srcCanvas.getContext("2d");
      srcCtx.drawImage(img, 0, 0);
      const imageData = srcCtx.getImageData(0, 0, srcCanvas.width, srcCanvas.height);

      const silhouette = detectSilhouette(imageData, srcCanvas.width, srcCanvas.height);
      if (!silhouette) {
        resolve({ normalized_url: imageUrl, normalization: null, silhouette: null });
        return;
      }

      const headRegion = detectHeadRegion(imageData, srcCanvas.width, srcCanvas.height, silhouette);
      const norm = computeNormalization(silhouette, headRegion, poseGeo, targetWidth);

      const outCanvas = document.createElement("canvas");
      outCanvas.width = targetWidth;
      outCanvas.height = targetHeight;
      const outCtx = outCanvas.getContext("2d");
      outCtx.fillStyle = "#FFFFFF";
      outCtx.fillRect(0, 0, targetWidth, targetHeight);

      if (norm) {
        const sw = silhouette.width_px;
        const sh = silhouette.height_px;
        const sx = silhouette.bounds.x * srcCanvas.width;
        const sy = silhouette.bounds.y * srcCanvas.height;
        const dw = sw * norm.scale;
        const dh = sh * norm.scale;
        const dx = norm.offset.x + (silhouette.bounds.x * targetWidth * norm.scale);
        const dy = norm.offset.y + (silhouette.bounds.y * targetHeight * norm.scale);

        outCtx.drawImage(srcCanvas, sx, sy, sw, sh, dx, dy, dw, dh);
      } else {
        const scale = Math.min(targetWidth * 0.6 / silhouette.width_px, targetHeight * 0.9 / silhouette.height_px);
        const dw = silhouette.width_px * scale;
        const dh = silhouette.height_px * scale;
        const dx = (targetWidth - dw) / 2;
        const dy = targetHeight * 0.95 - dh;
        outCtx.drawImage(srcCanvas, sx = silhouette.bounds.x * srcCanvas.width, silhouette.bounds.y * srcCanvas.height, silhouette.width_px, silhouette.height_px, dx, dy, dw, dh);
      }

      resolve({
        normalized_url: outCanvas.toDataURL("image/png"),
        normalization: norm,
        silhouette,
        head_region: headRegion,
      });
    };
    img.onerror = () => resolve({ normalized_url: imageUrl, normalization: null, silhouette: null });
    img.src = imageUrl;
  });
}

export function buildFaceMask(headRegion, canvasSize = 1024) {
  if (!headRegion) return null;
  const maskCanvas = document.createElement("canvas");
  maskCanvas.width = canvasSize;
  maskCanvas.height = canvasSize;
  const ctx = maskCanvas.getContext("2d");
  ctx.fillStyle = "#000000";
  ctx.fillRect(0, 0, canvasSize, canvasSize);

  const cx = headRegion.center.x * canvasSize;
  const cy = headRegion.center.y * canvasSize;
  const rx = (headRegion.bounds.w * canvasSize) / 2 * 1.2;
  const ry = (headRegion.bounds.h * canvasSize) / 2 * 1.1;

  ctx.fillStyle = "#FFFFFF";
  ctx.beginPath();
  ctx.ellipse(cx, cy, rx, ry, 0, 0, Math.PI * 2);
  ctx.fill();

  return maskCanvas.toDataURL("image/png");
}
